Breaking: Plugins envs and config removal (fixes #4782, fixes #4952)

nzakas opened the pull request:
This implements environments for plugins and also removes `rulesConfig` from plugins (because it was easier to do that than to fix all the failing tests for `rulesConfig` with this change).

ilyavolodin commented on lib/config.js:
```diff
@@ -15,7 +15,7 @@
     ConfigOps = require("./config/config-ops"),
     ConfigFile = require("./config/config-file"),
-    util = require("./util"),
+    Plugins = require("./config/plugins"),
```
Why capital "P"? Doesn't look like this is using `new` anywhere

nzakas replied:
This is the way I've been doing these in `./config`. See `ConfigFile`, `ConfigOps`, etc.

I was having a hard time distinguishing between object instances and decided to try and clarify by making them look more like classes with static methods.

ilyavolodin commented on lib/config/config-ops.js:
```diff
@@ -12,7 +12,7 @@
 
 var debug = require("debug"),
-    environments = require("../../conf/environments"),
+    Environments = require("./environments"),
```
Same as above. I though we only capitalize objects that need to be newed. We even have a rule for this.

nzakas commented:
Had to push to test. Once again, my local environment is borked for some reason (freezes on "Validating Rules"). I can't figure out what's going wrong, so I'll be pushing updates to validate fixes.

gyandeeps commented:
@nzakas thats happening on AppVeyor also. I had to cacel the run as it was running for 1 hr.
https://ci.appveyor.com/project/nzakas/eslint/build/139

nzakas commented:
Weird. Must be Windows-specific issue?

nzakas commented:
I'm able to get past that point on master, though, so something is different. :-/

ilyavolodin commented:
Tried this locally, and getting the same issue, so it's not just you.

gyandeeps commented:
On my machine also the execution freezes for unit test. But after a quick discetion of the problem, there are unit test failing for `cli-engine`. Most of them are where its not able to find a particular plugin or config module. 
Guess: My guess is that the error is getting silently eaten somewhere. Here is where the error comes up : https://github.com/eslint/eslint/blob/issue4782/lib/config/plugins.js#L84

nzakas commented:
Ah interesting. I'm still working on the errors, I'll see if that resolves the problem.

nzakas commented:
I think I've fixed the errors, so pushing that now. I still can't get `npm test` to run locally. :( Be on the lookout for anything on this diff that might be the source of the problem. I can run `npm test` on the current master without any problems, so it must be something I changed in this branch.

nzakas commented:
Still failing tests. :( Investigating.

nzakas commented:
With @gyandeeps's help, code coverage is where it needs to be. :) If anyone wants to take one last look at this before I merge, please do so.

ilyavolodin commented:
LGTM.
